src/nanotron/trainer.py

```diff
@@ -264,13 +261,35 @@ def post_training(self):
         pass
 
     def _print_training_plan(self):
-        stages_info = "".join(
-            f"[Stage {stage.name}] start from step {stage.start_training_step} \n" for stage in self.config.data_stages
-        )
-        full_log_message = f"[Training Plan] There are {len(self.config.data_stages)} training stages \n{stages_info}"
-        log_rank(full_log_message, logger=logger, level=logging.INFO, rank=0)
+        if hasattr(self.config, "data_stages") and self.config.data_stages is not None:
+            stages_info = "".join(
+                f"[Stage {stage.name}] start from step {stage.start_training_step} \n"
+                for stage in self.config.data_stages
+            )
+            full_log_message = (
+                f"[Training Plan] There are {len(self.config.data_stages)} training stages \n{stages_info}"
+            )
+            log_rank(full_log_message, logger=logger, level=logging.INFO, rank=0)
+
+    def _update_dataloader_based_on_training_stages(self, dataloaders: Union[List[DataLoader], DataLoader]):
+        from collections.abc import Generator
+
+        if not hasattr(self.config, "data_stages") or self.config.data_stages is None:
+            if self.current_dataloader is None:
+                if isinstance(dataloaders, tuple):
+                    dataloader = dataloaders[0]
+                else:
+                    dataloader = dataloaders
+                self.current_dataloader = sanity_check_dataloader(
+                    dataloader=dataloader, parallel_context=self.parallel_context, config=self.config
+                )
+            return
+        elif isinstance(dataloaders, Generator):
+            # TODO(xrsrke): this is a hacky way to handle DoReMi's dataloader
+            # remove this in the next PR
+            self.current_dataloader = dataloaders
```

Comment:
should we do sanity_check_dataloader here as well ?

Reply:
nah, it's already a generator